File: killrvideo-service-ratings/src/test/java/com/killrvideo/service/rating/grpc/RatingsServiceGrpcMapperTest.java
package com.killrvideo.service.rating.grpc;

import com.killrvideo.service.rating.dto.VideoRating;
import com.killrvideo.service.rating.dto.VideoRatingByUser;
import com.killrvideo.service.rating.request.GetUserRatingRequestData;
import killrvideo.ratings.RatingsServiceOuterClass.*;
import killrvideo.ratings.events.RatingsEvents;
import org.junit.jupiter.api.Test;

import java.util.UUID;

import static com.killrvideo.utils.GrpcMappingUtils.uuidToUuid;
import static org.junit.jupiter.api.Assertions.*;

class RatingsServiceGrpcMapperTest {
    private RatingsServiceGrpcMapper mapper = new RatingsServiceGrpcMapper();

    @Test
    public void testMapToRatingResponse() {
        UUID videoid = UUID.randomUUID();
        VideoRating vr = new VideoRating(videoid);
        GetRatingResponse proto = mapper.mapToRatingResponse(vr);
        assertEquals(vr.getVideoid().toString(), proto.getVideoId().getValue());
        assertEquals(0L, proto.getRatingsCount());
        assertEquals(0L, proto.getRatingsTotal());
    }

    @Test
    public void testMapToUserRatingResponse() {
        UUID videoid = UUID.randomUUID();
        UUID userid = UUID.randomUUID();
        VideoRatingByUser vr = new VideoRatingByUser(videoid, userid, 4);
        GetUserRatingResponse proto = mapper.mapToUserRatingResponse(vr);
        assertEquals(vr.getVideoid().toString(), proto.getVideoId().getValue());
        assertEquals(vr.getUserid().toString(), proto.getUserId().getValue());
        assertEquals(4, proto.getRating());
    }

    @Test
    public void testParseGetUserRatingRequest() {
        UUID videoid = UUID.randomUUID();
        UUID userid = UUID.randomUUID();
        GetUserRatingRequest request = GetUserRatingRequest.newBuilder()
                .setVideoId(uuidToUuid(videoid))
                .setUserId(uuidToUuid(userid))
                .build();

        GetUserRatingRequestData pojo = mapper.parseGetUserRatingRequest(request);
        assertEquals(videoid, pojo.getVideoid());
        assertEquals(userid, pojo.getUserid());
    }

    @Test
    public void testParseRateVideoRequest() {
        UUID videoid = UUID.randomUUID();
        UUID userid = UUID.randomUUID();
        RateVideoRequest request = RateVideoRequest.newBuilder()
                .setVideoId(uuidToUuid(videoid))
                .setUserId(uuidToUuid(userid))
                .build();

        VideoRatingByUser pojo = mapper.parseRateVideoRequest(request);
        assertEquals(videoid, pojo.getVideoid());
        assertEquals(userid, pojo.getUserid());
    }

    @Test
    public void testCreateUserRatedVideoEvent() {
        UUID videoid = UUID.randomUUID();
        UUID userid = UUID.randomUUID();
        VideoRatingByUser vr = new VideoRatingByUser(videoid, userid, 4);
        RatingsEvents.UserRatedVideo event = mapper.createUserRatedVideoEvent(vr);
        assertEquals(videoid.toString(), event.getVideoId().getValue());
        assertEquals(userid.toString(), event.getUserId().getValue());
        assertEquals(4, event.getRating());
        assertNotNull(event.getRatingTimestamp());
    }
}

File: killrvideo-service-ratings/src/test/kotlin/com/killrvideo/service/rating/grpc/RatingsServiceGrpcMapperTest.kt
package com.killrvideo.service.rating.grpc

import com.killrvideo.service.rating.dto.VideoRating
import com.killrvideo.service.rating.dto.VideoRatingByUser
import com.killrvideo.service.rating.grpc.RatingsServiceGrpcMapper.GetUserRatingRequestExtensions.parse
import com.killrvideo.service.rating.grpc.RatingsServiceGrpcMapper.RateVideoRequestExtensions.parse
import com.killrvideo.utils.GrpcMappingUtils.uuidToUuid
import killrvideo.ratings.getUserRatingRequest
import killrvideo.ratings.rateVideoRequest
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Test
import java.util.*

internal class RatingsServiceGrpcMapperTest {
    private val mapper = RatingsServiceGrpcMapper()
    @Test
    fun testMapToRatingResponse() {
        val videoid = UUID.randomUUID()
        val vr = VideoRating(videoid = videoid)
        val proto = mapper.mapToRatingResponse(vr)
        assertEquals(vr.videoid.toString(), proto.videoId.value)
        assertEquals(0L, proto.ratingsCount)
        assertEquals(0L, proto.ratingsTotal)
    }

    @Test
    fun testMapToUserRatingResponse() {
        val videoid = UUID.randomUUID()
        val userid = UUID.randomUUID()
        val vr = VideoRatingByUser(videoid = videoid, userid=userid, rating = 4)
        val proto = mapper.mapToUserRatingResponse(vr)
        assertEquals(vr.videoid.toString(), proto.videoId.value)
        assertEquals(vr.userid.toString(), proto.userId.value)
        assertEquals(4, proto.rating)
    }

    @Test
    fun testParseGetUserRatingRequest() {
        val videoid = UUID.randomUUID()
        val userid = UUID.randomUUID()
        val request = getUserRatingRequest {
            videoId = uuidToUuid(videoid)
            userId = uuidToUuid(userid)
        }
        val (videoid1, userid1) = request.parse()
        assertEquals(videoid, videoid1)
        assertEquals(userid, userid1)
    }

    @Test
    fun testParseRateVideoRequest() {
        val videoid = UUID.randomUUID()
        val userid = UUID.randomUUID()
        val request = rateVideoRequest {
            videoId = uuidToUuid(videoid)
            userId = uuidToUuid(userid)
        }
        val (videoid1, userid1) = request.parse()
        assertEquals(videoid, videoid1)
        assertEquals(userid, userid1)
    }

    @Test
    fun testCreateUserRatedVideoEvent() {
        val videoid = UUID.randomUUID()
        val userid = UUID.randomUUID()
        val vr = VideoRatingByUser(videoid = videoid, userid=userid, rating = 4)
        val event = mapper.createUserRatedVideoEvent(vr)
        assertEquals(videoid.toString(), event.videoId.value)
        assertEquals(userid.toString(), event.userId.value)
        assertEquals(4, event.rating)
        assertNotNull(event.ratingTimestamp)
    }
}
